lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -106,1360 +56,800 @@ function Install-Prerequisites {
-        if(!(Test-Path $pythonDir)) {
-            Throw "Could not find python directory"
-        }
+function Start-ConfigureVMSwitch {
+    $cfg = Get-JujuCharmConfig
+    $vmSwitchName = $cfg['vmswitch-name']
+    if (!$vmSwitchName) {
+        $vmSwitchName = $NOVA_DEFAULT_SWITCH_NAME
     }
-    return $pythonDir
-}
 
-function Generate-ExeWrappers {
-    if(!(Get-IsNanoServer)){
-        return
+    [array]$dataPorts, $managementOS = Get-DataPorts
+    $dataPort = $dataPorts[0]
+
+    $stateParams = @{
+        'Namespace' = "novahyperv"
+        'Key' = "vmswitch_name"
     }
-    $pythonDir = Get-PythonDir
-    $python = Join-Path $pythonDir "python.exe"
-    $updateWrapper = Join-Path $pythonDir "Scripts\UpdateWrappers.py"
 
-    $cmd = @($python, $updateWrapper, "nova-compute = nova.cmd.compute:main")
-    Invoke-JujuCommand -Command $cmd | Out-Null
+    $vmSwitches = [array](Get-VMSwitch -SwitchType External -ErrorAction SilentlyContinue)
+    foreach ($i in $vmSwitches) {
+        if ($i.NetAdapterInterfaceDescription -eq $dataPort.InterfaceDescription) {
+            $agentRestart = $false
 
-    $cmd = @($python, $updateWrapper, "neutron-hyperv-agent = neutron.cmd.eventlet.plugins.hyperv_neutron_agent:main")
-    Invoke-JujuCommand -Command $cmd
-}
+            if($i.Name -ne $vmSwitchName) {
+                $agentRestart = $true
+                Rename-VMSwitch $i -NewName $vmSwitchName | Out-Null
+                Set-CharmState @stateParams -Value $vmSwitchName
+            }
 
-function New-BondInterface {
-    if((Get-IsNanoServer)){
-        # not supported on nano yet
-        return $false
-    }
-    $name = Get-JujuCharmConfig -Scope "bond-name"
-    $bondPorts = Get-InterfaceFromConfig -ConfigOption "bond-ports"
-    if ($bondPorts.Length -eq 0) {
-        return $false
-    }
+            if($i.AllowManagementOS -ne $managementOS) {
+                $agentRestart = $true
+                Set-VMSwitch -Name $vmSwitchName -AllowManagementOS $managementOS | Out-Null
+            }
 
-    $bondExists = Get-NetLbfoTeam -Name $name -ErrorAction SilentlyContinue
-    if ($bondExists){
-        return $true
+            if($agentRestart) {
+                $netType = Get-NetType
+                if($netType -eq "ovs") {
+                    $status = (Get-Service -Name "ovs-vswitchd").Status
+                    if($status -eq [System.ServiceProcess.ServiceControllerStatus]::Running) {
+                        Restart-Service "ovs-vswitchd" | Out-Null
+                    }
+                }
+            }
+            return
+        }
     }
 
-    $bond = New-NetLbfoTeam -Name $name -TeamMembers $bondPorts.Name -TeamNicName $name -TeamingMode LACP -Confirm:$false
-    $isUp = WaitFor-BondUp -bond $bond.Name
-    if (!$isUp){
-        Throw "Failed to bring up $name"
+    $oldSwitchName = Get-CharmState @stateParams
+    if($oldSwitchName) {
+        Remove-VMSwitch -Name $oldSwitchName -Force -Confirm:$false
```

Comment:
nova-hyperv charm useses `novahyperv` namespace from registry to save data. And when the charm is removed from the machine it also remove everything from this namespace.

Do you see any scenario when this will fail ? I assumed no manual intervention or no other charms using the same namespace as nova-hyperv charm.